src/main/scala/mesosphere/marathon/scheduling/LegacyScheduler.scala

```diff
@@ -0,0 +1,116 @@
+package mesosphere.marathon
+package scheduling
+
+import akka.Done
+import mesosphere.marathon.core.instance.update.InstanceUpdateOperation.RescheduleReserved
+import mesosphere.marathon.core.instance.{Goal, Instance}
+import mesosphere.marathon.core.launcher.OfferProcessor
+import mesosphere.marathon.core.launchqueue.LaunchQueue
+import mesosphere.marathon.core.task.termination.{KillReason, KillService}
+import mesosphere.marathon.core.task.tracker.InstanceTracker
+import mesosphere.marathon.core.task.update.TaskStatusUpdateProcessor
+import mesosphere.marathon.state.{PathId, RunSpec}
+import mesosphere.marathon.util.WorkQueue
+import org.apache.mesos.Protos
+
+import scala.async.Async.{async, await}
+import scala.concurrent.{ExecutionContext, Future}
+
+case class LegacyScheduler(
+    offerProcessor: OfferProcessor,
+    instanceTracker: InstanceTracker,
+    statusUpdateProcessor: TaskStatusUpdateProcessor,
+    killService: KillService,
+    launchQueue: LaunchQueue) extends Scheduler {
+
+  // We cannot process more Add/Update requests for one runSpec in parallel because it leads to race condition.
+  // See MARATHON-8320 for details. The queue handling is helping us ensure we change one instance at a time.
+  private[this] val serializeUpdates: WorkQueue = WorkQueue("LegacyScheduler", maxConcurrent = 1, maxQueueLength = Int.MaxValue)
+
+  override def schedule(runSpec: RunSpec, count: Int)(implicit ec: ExecutionContext): Future[Done] = serializeUpdates {
+    async {
+      val instancesToSchedule = 0.until(count).map { _ => Instance.scheduled(runSpec, Instance.Id.forRunSpec(runSpec.id)) }
+      await(instanceTracker.schedule(instancesToSchedule))
+
+      // Let launch queue manage task launcher actor.
+      await(launchQueue.sync(runSpec))
+
+      Done
+    }
+  }
+
+  override def reschedule(instance: Instance, runSpec: RunSpec)(implicit ec: ExecutionContext): Future[Done] = serializeUpdates {
```

Comment:
as I wrote in other comments, I would actually remove this method, I think it's an implementation detail

Reply:
It is *not* an implementation detail. There is a semantic difference between `reschedule` aka `update` and `schedule` aka `create`.